apps/emqx_durable_storage/src/emqx_ds_storage_bitfield_lts.erl

```diff
@@ -0,0 +1,382 @@
+        ts_bits := TSBits,
+        ts_offset_bits := TSOffsetBits
+    } = Schema,
+    {_, DataCF} = lists:keyfind(data_cf(GenId), 1, CFRefs),
+    {_, TrieCF} = lists:keyfind(trie_cf(GenId), 1, CFRefs),
+    Trie = restore_trie(TopicIndexBytes, DBHandle, TrieCF),
+    %% If user's topics have more than learned 10 wildcard levels
+    %% (more than 2, really), then it's total carnage; learned topic
+    %% structure won't help.
+    MaxWildcardLevels = 10,
+    KeymapperCache = array:from_list(
+        [
+            make_keymapper(TopicIndexBytes, BitsPerTopicLevel, TSBits, TSOffsetBits, N)
+         || N <- lists:seq(0, MaxWildcardLevels)
+        ]
+    ),
+    #s{db = DBHandle, data = DataCF, trie = Trie, keymappers = KeymapperCache}.
+
+-spec store_batch(
+    emqx_ds_replication_layer:shard_id(), s(), [emqx_types:message()], emqx_ds:message_store_opts()
+) ->
+    emqx_ds:store_batch_result().
+store_batch(_ShardId, S = #s{db = DB, data = Data}, Messages, _Options) ->
+    lists:foreach(
+        fun(Msg) ->
+            {Key, _} = make_key(S, Msg),
+            Val = serialize(Msg),
+            rocksdb:put(DB, Data, Key, Val, [])
+        end,
+        Messages
+    ).
+
+get_streams(_Shard, #s{trie = Trie}, TopicFilter, _StartTime) ->
+    Indexes = emqx_ds_lts:match_topics(Trie, TopicFilter),
+    [#stream{storage_key = I} || I <- Indexes].
+
+make_iterator(_Shard, _Data, #stream{storage_key = StorageKey}, TopicFilter, StartTime) ->
+    %% Note: it's a good idea to keep the iterator structure lean,
+    %% since it can be stored on a remote node that could update its
+    %% code independently from us.
+    {ok, #it{
+        topic_filter = TopicFilter,
+        start_time = StartTime,
+        storage_key = StorageKey
+    }}.
+
+next(_Shard, #s{db = DB, data = CF, keymappers = Keymappers}, It0, BatchSize) ->
+    #it{
+        start_time = StartTime,
+        storage_key = StorageKey
+    } = It0,
+    %% Make filter:
+    {TopicIndex, Varying} = StorageKey,
+    Inequations = [
+        {'=', TopicIndex},
+        {'>=', StartTime}
+        | lists:map(
+            fun
+                ('+') ->
+                    any;
+                (TopicLevel) when is_binary(TopicLevel) ->
+                    {'=', hash_topic_level(TopicLevel)}
+            end,
+            Varying
+        )
+    ],
+    %% Obtain a keymapper for the current number of varying
+    %% levels. Magic constant 2: we have two extra dimensions of topic
+    %% index and time; the rest of dimensions are varying levels.
+    NVarying = length(Inequations) - 2,
+    Keymapper = array:get(NVarying, Keymappers),
+    Filter =
+        #filter{range_min = LowerBound, range_max = UpperBound} = emqx_ds_bitmask_keymapper:make_filter(
+            Keymapper, Inequations
+        ),
+    {ok, ITHandle} = rocksdb:iterator(DB, CF, [
+        {iterate_lower_bound, emqx_ds_bitmask_keymapper:key_to_bitstring(Keymapper, LowerBound)},
```

Comment:
Probably it's not going to improve much, since we never go backwards. Moreover, when the iterator is first created, `last_seen_key` is `<<>>`.

Reply:
> since we never go backwards

Q: I thought the idea was to shrink iteration range to give RocksDB a chance to exclude as much SSTs as possible (with the help of bloom filters and whatnot). Why does going backward matter here?

Reply:
I didn't attempt to optimize performance in this particular PR, because it's no use before benchmarking is in place. (I was trying to avoid pessimizations, though). 

The reason I introduced upper and lower bounds was to avoid dealing with keys that have different structure (e.g. more or less varying levels).